url46a88cc932_toby1984_tinyc_tgz-pJ8-de_codesourcery_dcpu16_compiler_CompilerJ8/src/de/codesourcery/dcpu16/parser/ShuntingYard.java

```diff
@@ -0,0 +1,316 @@
+    private final Stack<ExpressionToken> stack = new Stack<ExpressionToken>();
+
+    private final Stack<Integer> argsCountStack = new Stack<Integer>();
+    private final Stack<Boolean> argsMarkerStack = new Stack<Boolean>();       
+    
+    public ShuntingYard() {
+    }
+    
+    public static final class MismatchedClosingParensException extends ParseException {
+
+        public MismatchedClosingParensException(ITextRegion r)
+        {
+            super("Mismatched closing parens", r);
+        }
+    }
+    
+    public boolean isFunctionOnStack() 
+    {
+        for ( ExpressionToken t : stack ) {
+            if ( t.isFunction() ) {
+                return true;
+            }
+        }
+        return false;
+    }
+    
+    public boolean isEmpty() {
+        return valueQueue.isEmpty() && stack.isEmpty();
+    }
+
+    public void pushValue(ASTNode node) 
+    {
+        output( new ExpressionToken(ExpressionTokenType.VALUE , node ) );
+        // If the were values stack has a value on it, pop it and push true
+        if ( ! argsMarkerStack.isEmpty() ) {
+            argsMarkerStack.pop();
+            argsMarkerStack.push( Boolean.TRUE );
+        }
+    }        
+
+    /*
+    [X] Read a token.
+    [X] If the token is a number, then add it to the output queue.  
+    [X] If the token is a function token, then push it onto the stack. 
+    [X] If the token is a left parenthesis, then push it onto the stack.
+
+    [X] If the token is a function argument separator (e.g., a comma):
+
+        Until the token at the top of the stack is a left parenthesis, pop operators off the stack onto the output queue. If no left parentheses are encountered, either the separator was misplaced or parentheses were mismatched.
+
+    [X] If the token is an operator, o1, then:
+
+        while there is an operator token, o2, at the top of the stack, and
+
+                either o1 is left-associative and its precedence is less than or equal to that of o2,
+                or o1 has precedence less than that of o2,
+
+            pop o2 off the stack, onto the output queue;
+
+        push o1 onto the stack.
+
+
+    [X] If the token is a right parenthesis:
+
+        Until the token at the top of the stack is a left parenthesis, pop operators off the stack onto the output queue.
+        Pop the left parenthesis from the stack, but not onto the output queue.
+        If the token at the top of the stack is a function token, pop it onto the output queue.
+        If the stack runs out without finding a left parenthesis, then there are mismatched parentheses.
+
+When there are no more tokens to read:
+
+    While there are still operator tokens in the stack:
+
+        If the operator token on the top of the stack is a parenthesis, then there are mismatched parentheses.
+        Pop the operator onto the output queue.
+
+     */
+
+    public void pushOperator(ExpressionToken tok1) 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 113-200 into a separate method.